File: guests/go/entry/entry.go
// The entry scene (tools/scenes/entry.steps). THE OTHER EVENT SURFACE: the
// handlers register on the APP after the build, not on the constructors.
package entry

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

func App() *kaya.App {
	app := kaya.NewApp()

	var (
		status     kaya.Signal[string]
		field, add kaya.Widget
		todos      kaya.Collection
	)

	app.Build(func(tx *kaya.Tx) {
		status = tx.Signal("no todos")
		todos = tx.Collection()

		tx.Mount(tx.Column(func() {
			field = tx.Entry(nil)
			add = tx.Button("add", nil)
			tx.Label(status)
			for row := range tx.Rows(todos).All() {
				row.Label(row.Value())
			}
		}))
	})

	draft := ""
	app.OnChange(field, func(tx *kaya.Tx, text string) {
		draft = text
	})
	app.OnClick(add, func(tx *kaya.Tx) {
		if draft == "" {
			total := tx.Len(todos)
			tx.Write(status, fmt.Sprintf("nothing to add, %d total", total))
			return
		}
		// NO KEY: the binding mints the name and hands it back
		// (docs/fresh-key-plan.md).
		tx.InsertFresh(todos, draft)
		total := tx.Len(todos)
		tx.Write(status, fmt.Sprintf("added %s, %d total", draft, total))
		tx.Clear(field)
		tx.Focus(field)
	})

	return app
}

File: guests/java/dev/kaya/guests/Entry.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

/**
 * The entry scene from the JVM — guests/rust/entry.rs, tools/scenes/entry.steps.
 */
public final class Entry {
    private static final class Scene {
        final KayaApp.Signal<String> status;
        final KayaApp.Widget field;
        final KayaApp.Widget add;
        final KayaApp.Collection todos;

        Scene(KayaApp.Signal<String> status, KayaApp.Widget field, KayaApp.Widget add,
                KayaApp.Collection todos) {
            this.status = status;
            this.field = field;
            this.add = add;
            this.todos = todos;
        }
    }

    /** Java lambdas cannot assign captured locals. */
    private static final class Refs {
        KayaApp.Widget field, add;
    }

    private static String draft = "";

    public static void app() {
        KayaApp app = new KayaApp();

        Scene scene = app.build(tx -> {
            KayaApp.Signal<String> status = tx.signal("no todos");
            KayaApp.Collection todos = tx.collection();

            Refs refs = new Refs();
            tx.mount(tx.column(() -> {
                refs.field = tx.entry(); // entry#0
                refs.add = tx.button("add"); // button#0
                tx.label(status); // label#0
                for (var row : tx.rows(todos)) {
                    row.label(row.value());
                }
            }));
            return new Scene(status, refs.field, refs.add, todos);
        });

        app.onChange(scene.field, (tx, text) -> draft = text);
        app.onClick(scene.add, tx -> {
            if (draft.isEmpty()) {
                tx.write(scene.status, "nothing to add, " + tx.count(scene.todos) + " total");
                return;
            }
            // The binding mints the key (docs/fresh-key-plan.md).
            tx.insertFresh(scene.todos, draft);
            int total = tx.count(scene.todos);
            tx.write(scene.status, "added " + draft + ", " + total + " total");
            // The clear comes back as text_changed(""), so onChange empties draft.
            tx.clear(scene.field);
            tx.focus(scene.field);
        });

        app.dispatchLoop();
    }

    private Entry() {}
}
